src/parser.rs

```diff
@@ -1311,20 +1324,45 @@ impl Library {
             _ => Err(parser.unexpected_element(elem)),
         })?;
 
-        if let Some((tid, c_type, mut array_length)) = typ {
-            if for_method {
-                array_length = array_length.map(|l| l + 1);
+        if let Some(mut elem_type) = typ {
+            if elem_type.is_array {
+                // Note: these could be handled in the `analysis` part, however that would be quite
+                //       annoying because many code places check the parsed `nullable` instead of
+                //       the `analysis` result. Gir necessitates a rewrite...
+                use ParameterDirection::*;
+                match direction {
+                    Return | Out | InOut => {
+                        // Some arrays in return position come with nullable=true & zero-terminated=false
+                        // annotations. In practice, the C implementations return NULL when the size is 0.
```

Comment:
You mean with `nullable=false` and `zero-terminated=false`? That's actually an interesting case because not returning NULL in that situation is actually hard for C code (`malloc(0)` is implementation defined behaviour IIRC).

For `nullable=false` and `zero-terminated=true` it's IMHO a bug in the C code and it should return a pointer that dereferences to `NULL`.

Reply:
* In parameters: `Option` if nullable, otherwise not
* Return/out parameters: `Option` if nullable and zero-terminated, no `Option` if nullable and not zero-terminated (and obviously no `Option` if not nullable)